### PR TITLE
feat(replay): separate feedback from attachments in the same envelope

### PR DESCRIPTION
**Pull request by aliu39**
The idea is to restructure the code so that UserReportV2's don't end up in the catch-all extract_kafka_messages at the bottom of `store_envelope`, with attachments

Relates to https://github.com/getsentry/team-replay/issues/393

### CODE REVIEW

File: relay-server/src/services/store.rs
Comment (Dav1dde) [inline relay-server/src/services/store.rs]: It's better if you pass in the topic instead or even better, let the produce function figure out the topic.

File: relay-server/src/services/store.rs
Comment (Dav1dde) [inline relay-server/src/services/store.rs]: Unused with a bunch of todos?

File: relay-server/src/services/store.rs
Comment (Dav1dde) [inline relay-server/src/services/store.rs]: That's something which is better tracked in a ticket than code.

**Review by Dav1dde — COMMENTED** [review]
Please see if you can add some tests or modify existing tests (if there are any, why didn't they break with that change?).

Also before moving on with this to SaaS it'd be best if you figured out the wire format for the Feedback topic, it's better to avoid supporting multiple formats on the other "end".

**Comment by aliu39** [comment]
> Also before moving on with this to SaaS it'd be best if you figured out the wire format for the Feedback topic, it's better to avoid supporting multiple formats on the other "end".

Thanks for the review! Could you elaborate what you mean by "wire format"? Do you mean the `UserReportV2KafkaMessage` type?

You're right, I'll make a ticket for adding that message type and remove it from this PR, it's something I forgot to clean that up. After adding coverage I could deploy this with a FF enabled for s4s only.

**Comment by Dav1dde** [comment]
> Thanks for the review! Could you elaborate what you mean by "wire format"? Do you mean the UserReportV2KafkaMessage type?

Yes, you're using the `Event` right now and it looks like you want to move to `UserReportV2KafkaMessage`, it's better if you do it early, it's gonna be a lot harder to do later if you have to support both versions (we can't instantly switch Relays to a new version so for a certain period of time all downstream consumers have to support both formats).

**Comment by aliu39** [comment]
> Yes, you're using the `Event` right now and it looks like you want to move to `UserReportV2KafkaMessage`, it's better if you do it early, it's gonna be a lot harder to do later if you have to support both versions (we can't instantly switch Relays to a new version so for a certain period of time all downstream consumers have to support both formats).

Got it, I'll definitely look into it and give it more thought before finalizing this PR. Thanks for the suggestion!

**Comment by aliu3ntry** [comment]
Removing commented code in a sec!

**Comment by aliu39** [comment]
> Also before moving on with this to SaaS it'd be best if you figured out the wire format for the Feedback topic, it's better to avoid supporting multiple formats on the other "end".

I've decided to hold off on making a UserReportV2KafkaMessage type until the new topic/infra has been deployed to all regions in prod. It adds an extra layer of complexity and testing that will delay this feature, which the SDK team needs to GA user feedback in ~1 week.

It probably shouldn't be done until we write a separate ingest strategy for the feedback consumer (right now it shares this with events).

File: relay-server/src/services/store.rs
Comment (JoshFerge) [inline relay-server/src/services/store.rs]: i think these comments could be better placed in the ItemType definition?

File: relay-server/src/services/store.rs
Comment (JoshFerge) [inline relay-server/src/services/store.rs]: not sure what this naming implies, maybe just `separate_attachments_from_feedback`?

**Review by JoshFerge — COMMENTED** [review]
other than the small nits / naming, lgtm

File: relay-server/src/services/store.rs
Comment (aliu39) [inline relay-server/src/services/store.rs:702]: See https://github.com/getsentry/relay/pull/3344

File: relay-server/src/services/store.rs
Comment (aliu39) [reply relay-server/src/services/store.rs]: Decided to name it this because it also results in some code restructuring / different code path for producing feedback in StoreService.

I agree the naming's a little funky though, will sleep on it + see what Vienna thinks?

Comment (jjbayer) [reply relay-server/src/services/store.rs]: I would also prefer aligning the variable name (and the field name `feedback_ingest_v2`) with either the feature name ("inline attachments") or renaming it to describe what it actually enables.

File: relay-server/src/services/store.rs
Comment (jjbayer) [inline relay-server/src/services/store.rs]: With this, you can make sure that there's an `error` log if there's an unexpected feedback item in the envelope.

```suggestion
                ItemType::UserReportV2 if use_feedback_ingest_v2 => {
```

File: relay-server/src/services/store.rs
Comment (jjbayer) [inline relay-server/src/services/store.rs:710]: Out of scope for this PR, but should we hard-code this to `Feedback` soon? Or are some installations not ready for that yet? (self-hosted, devenv, ...)

File: tests/.idea/workspace.xml
Comment (jjbayer) [inline tests/.idea/workspace.xml:1]: Did you check this in on purpose?

File: tests/integration/test_feedback.py
Comment (iker-barriocanal) [inline tests/integration/test_feedback.py]: Why the `_` suffix?

File: relay-server/src/services/store.rs
Comment (iker-barriocanal) [reply relay-server/src/services/store.rs]: The flag's docs say it's temporary. Is it only used for testing in development or should we update the docs?

Comment (aliu39) [reply relay-server/src/services/store.rs]: It's used like a feature flag - we'll set it to true in test region(s), test the SDK, and remove this logic + non-v2 code path once satisfied. The flag is just to quickly "revert" in case things start to break

Comment (aliu39) [reply relay-server/src/services/store.rs]: What did you have in mind for updating the docs?

File: relay-server/src/services/store.rs
Comment (aliu39) [reply relay-server/src/services/store.rs:710]: Yes this has to do with ops deployment of the feedback topic -- right now, its only in s4s and not other regions. This check and rollout rate will eventually be removed too

File: tests/.idea/workspace.xml
Comment (aliu39) [reply tests/.idea/workspace.xml:1]: No 😅

Comment (iker-barriocanal) [inline tests/.idea/workspace.xml:1]: Please, remove this file.

File: relay-dynamic-config/src/global.rs
Comment (iker-barriocanal) [inline relay-dynamic-config/src/global.rs:171]: Without context, this comment is not clear to me what the flag should do. What do you think about something like the following?

```suggestion
    /// Kill switch to stop producing feedback items separately into Kafka. The topic is controlled by `feedback_ingest_topic_rollout_rate`.
```

File: tests/integration/test_feedback.py
Comment (iker-barriocanal) [inline tests/integration/test_feedback.py:174]: This option is overwritten on either line 179 or 183, let's remove this one.

File: tests/integration/test_feedback.py
Comment (iker-barriocanal) [inline tests/integration/test_feedback.py:186]: Unless there's a reason to have it:
```suggestion
        feedback_consumer = events_consumer(timeout=20)
        other_consumer = feedback_consumer(timeout=20)
    attachments_consumer = attachments_consumer(timeout=20)
```
(same with the `if` block)

**Review by iker-barriocanal — APPROVED** [review]
LGTM! Please see comments below before merging.

**Comment by iker-barriocanal** [comment]
@aliu3ntry please address the suggestions above, I didn't realize the PR had auto-merging enabled.

File: tests/integration/test_feedback.py
Comment (aliu3ntry) [reply tests/integration/test_feedback.py:186]: Sounds good, I didn't want to shadow/override the fixtures with the same name, but should be fine for this test!

Comment (aliu39) [reply tests/integration/test_feedback.py:186]: Sounds good, I didn't want to shadow/override the fixtures with the same name, but can make this change!

File: tests/integration/test_feedback.py
Comment (aliu39) [reply tests/integration/test_feedback.py:174]: good catch!

**Comment by aliu39** [comment]
Oops, sorry! Reopened at https://github.com/getsentry/relay/pull/3486 @iker-barriocanal

**Pull request merged**
